rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,151 @@
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+[ inherit (attrs) a b c; ]
+[ inherit a b c; ]
+```
+
+These expressions are (respectively) syntactic sugar for:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+[ a b c ]
+```
+
+The `inherit` keyword is intentionally reused so as to not add any new
+keywords to the Nix grammar, and to evoke the similarities with the
+existing attribute-based inherit syntax.
+
+As the `inherit` keyword is currently a syntax error inside a list expression,
+a Nix interpreter which supports this new language feature will be compatible
+with existing Nix code.
+
+This RFC is implemented here: https://github.com/nixos/nix/pull/5402
+
+For MVP functionality, only minor additions to `src/libexpr/parser.y` are
+needed. If accepted, the changes to the Nix interpreter can be backported
+to current releases if desired. Relevant language documentation and
+third-party parsers/linters would also need to be updated.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This would be useful for many languages and frameworks in Nixpkgs which
+extract packages from a package set argument.
+
+For example, `python3.withPackages (ps: [ inherit (ps) ...; ])` will serve as a
+more fine-grained alternative to `python3.withPackages (ps: with ps; [ ... ])`.
+This would apply similarly to `vim.withPlugins`, `lua.withPackages`, etc.
+
+Certain list-typed `meta` fields could also make use of this feature, e.g.:
+```
+meta.licenses = [ inherit (lib.licenses) bsd3 mit; ];
+meta.maintainers = [ inherit (lib.maintainers) johndoe janedoe; ];
+```
+
+List-inherits can be used multiple times in a list-expression, and intermingled
+with ordinary list elements without needing to concatenate:
+
+```
+buildInputs = [
+  a b c /* ordinary list elements */
+  inherit (pkgsFoo) d e f;
+  g h i /* ordinary list elements */
+  inherit (pkgsBar) j k l;
+];
+```
+
+The order of the elements is preserved inside of and between inherits,
+so the previous is exactly equivalent to
+
+```
+buildInputs = [
+  a b c
+  pkgsFoo.d pkgsFoo.e pkgsFoo.f
+  g h i
+  pkgsBar.j pkgsBar.k pkgsBar.l
+];
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This will add complexity to the Nix grammar and any third-party tools which
+  operate on Nix expressions.
+* Expressions reliant on the new syntax will be incompatible with
+  Nix versions prior to the introduction of this feature.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* Skillful use of `with` to avoid its drawbacks
+* Fine-grained attribute selection via existing (more verbose) language
+  features, such as `builtins.attrValues { inherit (attrs) a b c; }`
```

Comment:
This alternative has one downside, since `attrValues` returns a alphabetically sorted list. Repl example:
```
> attrs = { a=1; b=2; }
> builtins.attrValues { inherit (attrs) b a; }
[1 2]
```
but `[2 1]` was specified.

Reply:
Yeah, definitely unintuitive. I think the worst aspect is that renaming an attr may reorder the resulting list despite not changing any values:

```
nix-repl> builtins.attrValues { a = 1; b = 2; }        
[ 1 2 ]

nix-repl> builtins.attrValues { c = 1; b = 2; } 
[ 2 1 ]
```